### PR TITLE
step-2: Make program more useful.

### PR DESCRIPTION
**Pull request by bangerth**
My students in my deal.II course asked me what the point is of showing the sparsity pattern as the only real outcome of step-2. I didn't have a good reason.

The point of step-2 is to illustrate the enumeration of degrees of freedom. The sparsity pattern is really just the next step before we can actually solve equations, but we should show something about the outcome of the first step, namely enumerating degrees of freedom. This patch adds this by outputting DoF locations and describing the resulting output.

Opinions welcome.

### CODE REVIEW

**Review by kronbichler — APPROVED** [review]
This looks great to me.

**Comment by bangerth** [comment]
So fixed.

**Comment by bangerth** [comment]
For reference, the new pictures are here:
https://dealii.org/images/steps/developer/step-2.dof_locations_1.png
https://dealii.org/images/steps/developer/step-2.dof_locations_2.png

Or, by way of just showing them here:
image
image

**Comment by tjhei** [comment]
I like the idea, but the overlapping DoF labels in the pictures are not exactly pretty. Is there a way to reduce the text size and also move them closer to the actual DoF location?

**Comment by bangerth** [comment]
Yes, true. The program uses the same mesh as step-1, and I think that's a useful feature because we try to keep as many things the same from one program to another in the early tutorials.

What if I commented on the fact that this output already illustrates that this kind of output is not particularly useful, for exactly the reason you state? I do say in the program that the exact location of vertices is almost always not something you're interested in, and this plot also illustrates why.

In other words, would you be ok if I addressed the issue with text?

**Comment by tjhei** [comment]
One other thought (also see the conversation on slack about it):
While it is neat to show sparsity patterns and DoF locations, it might not be something that the user needs to learn this early (in their second tutorial program).
Another thought would be to interpolate and visualize a finite element function and maybe do a point evaluation. These are all fundamental concepts that are important.

**Comment by bangerth** [comment]
I'm amenable to all of these ideas, but as you know I'm an incrementalist. If you tell me what I need to do to get this PR merged, I'm happy to do other steps at a later time.

**Comment by bangerth** [comment]
Ping?

**Comment by tjhei** [comment]
I am still not convinced that this is a great choice for step-2 especially if we add other stuff (interpolate and visualize a solution).

Any other votes for dof locations (this PR) vs interpolating a function vs both?

**Comment by blaisb** [comment]
I would personally vote for interpolating a function in 1d and 2D, but that's my opinion and I am comfortable with many things. However, I think showing the DOF location AND interpolating is also very compatible, because the DOF locations are the points where you would evaluate the function you are interpolating. So to me, both in the same step would not be too many lines. The thing I would remove is the whole idea of the sparsity pattern, because we are not solving any matrices at that point.

**Comment by bangerth** [comment]
Compromise suggestion: We merge this one, showing DoF locations, and I add interpolation onto the space in a separate patch that would then also remove the sparsity pattern stuff.

**Comment by bangerth** [comment]
On 3/6/23 05:14, Timo Heister wrote:
> Sounds good. Can we fit the sparsity pattern visualization somewhere else? 
> Possibilities for extension of step-2?

Yes, that's a nice idea. I will do so in my follow-up.

@blais: Let's keep things simple. One thing I learned over the years that from 
an educational perspective, things work best if they stay simple. Just because 
something *can* be done doesn't mean that it *should* be done.

**Review by drwells — APPROVED** [review]
Looks good.

It looks like

https://www.dealii.org/images/steps/developer/step-2.dof-locations-1.svg

are not yet online - could we add them before merging.

**Comment by blais** [comment]
Wrong user, I think you meant Bruno (@blaisb).
(Nice library!)

**Comment by drwells** [comment]
Small world... a friend of mine started using beancount a few weeks ago and loves it!

**Comment by bangerth** [comment]
I had the files online already with underscores instead of dashes. Fixed now -- the patch is ready to go!

**Comment by drwells** [comment]
The files you uploaded are PNGs but their file extensions are `.svg` - can you go in and change the file extensions?

**Comment by bangerth** [comment]
On 3/25/23 18:24, David Wells wrote:
> The files you uploaded are PNGs but their file extensions are |.svg| - can you 
> go in and change the file extensions?

Ah, can't get this right apparently! Should be fixed now!

File: examples/step-2/doc/results.dox [hunk 1]
```diff
@@ -1,14 +1,43 @@
...
+```
+plot [-.2:2.2][-1.2:1.2] "mesh.gnuplot" with lines, "dof_locations_2.gnuplot" using 1:2:3 with labels point offset 1,1
+```
+This may be cryptic, but it is also not particularly important and you shouldn't
+spend too much time understanding what it does. Rather, the important part is
+to look at what we get as output:
+<table style="width:60%" align="center">
+  <tr>
+    <td><img src="https://www.dealii.org/images/steps/developer/step-2.dof-locations-1.svg" alt=""></td>
+    <td><img src="https://www.dealii.org/images/steps/developer/step-2.dof-locations-2.svg" alt=""></td>
```
Comment (drwells) [inline examples/step-2/doc/results.dox:22]:
```suggestion
    <td><img src="https://www.dealii.org/images/steps/developer/step-2.dof-locations-1.png" alt=""></td>
    <td><img src="https://www.dealii.org/images/steps/developer/step-2.dof-locations-2.png" alt=""></td>
```
Almost - the files are now correct but the links in the tutorial still point to the (now nonexistent) SVG files.

**Comment by tjhei** [comment]
I think the numbering in .png pictures is barely readable. Can you try changing the offset from the point location and/or use a bigger figure size so that the numbers are smaller relative to the mesh?

**Comment by bangerth** [comment]
I updated the font size in the pictures. This now looks as follows: 
image

**Comment by bangerth** [comment]
Anyone?

**Comment by tjhei** [comment]
> I updated the font size in the pictures. This now looks as follows:

This is certainly better. Do you have the precise instructions to generate this figure? Can we maybe put this somewhere (maybe next to the image)? It seems that you are no longer setting the aspect ratio correctly.

File: examples/step-2/doc/intro.dox [hunk 2]
```diff
@@ -53,6 +76,23 @@ live here". This is the sort of information you need when determining
 how big your system matrix should be, and when copying the
 contributions of a single cell into the global matrix.
 
+The first task of the current program is therefore to take a mesh
+and a finite element, and simply enumerate the degrees of freedom. In the
```
Comment (tjhei) [inline examples/step-2/doc/intro.dox:80]:
```suggestion
and a finite element, and enumerate the degrees of freedom. In the
```

File: examples/step-2/doc/intro.dox [hunk 3]
```diff
@@ -53,6 +76,23 @@ live here". This is the sort of information you need when determining
...
+current context, this means simply giving each vertex of the mesh a
+DoF index. Once that has happened, we will output in a picture which
+vertex ended up with which DoF index. You can find the corresponding
+pictures in the <a href="#Results">results section</a> of this program.
```
Comment (tjhei) [inline examples/step-2/doc/intro.dox:84]:
```suggestion
pictures in the <a href="#Results">results section</a> of this tutorial.
```

File: examples/step-2/doc/intro.dox [hunk 4]
```diff
@@ -53,6 +76,23 @@ live here". This is the sort of information you need when determining
...
+
+It is probably worth pointing out that where each DoF is geometrically
+located is not actually a question we typically ask in finite element
```
Comment (tjhei) [inline examples/step-2/doc/intro.dox:87]:
```suggestion
located is not a question we typically ask in finite element
```

File: examples/step-2/doc/intro.dox [hunk 5]
```diff
@@ -53,6 +76,23 @@ live here". This is the sort of information you need when determining
...
+codes. Most often, we only care about the fact that there *is* an
+enumeration of all degrees of freedom, but which DoF concretely is
+where is not important to us. (We will also come back to this below
```
Comment (tjhei) [inline examples/step-2/doc/intro.dox:90]:
```suggestion
codes. Most often, we only care about the fact that there *is* an
enumeration of all degrees of freedom and not where each DoF
is located at. (We will also come back to this below
```

File: examples/step-2/step-2.cc [hunk 6]
```diff
@@ -58,11 +62,19 @@
...
+
 // @sect3{Mesh generation}
 
 // This is the function that produced the circular grid in the previous step-1
 // example program with fewer refinements steps. The sole difference is that it
 // returns the grid it produces via its argument.
+//
+// At the end of the function, we also output this mesh into a file. We will
+// use this as one piece of information when visualizing which degree of
+// freedom is where. To output a mesh, we use the GridOut class that you have
```
Comment (tjhei) [inline examples/step-2/step-2.cc:74]:
```suggestion
// use this as one piece of information when visualizing the location of degrees of
// freedom. To output a mesh, we use the GridOut class that you have
```

File: examples/step-2/step-2.cc [hunk 7]
```diff
@@ -126,12 +181,15 @@ void distribute_dofs(DoFHandler<2> &dof_handler)
...
-  // factors, of which we will demonstrate one in the following.
+  // Now that we have associated a degree of freedom with a global
+  // number to each vertex, Let us output this information using the
+  // function above:
+  write_dof_locations(dof_handler, "dof-locations-1.gnuplot");
+
+  // In practice, we do not often care about where a degree of freedom
+  // is geometrically located, and so other than seeing it once via
+  // the call above is not practically useful. But where two degrees
+  // of freedom in relation to each other matters in other ways.
```
Comment (tjhei) [inline examples/step-2/step-2.cc:192]:
```suggestion
  // of freedom are in relation to each other matters in other ways.
```
?

**Review by tjhei — APPROVED** [review]

File: examples/step-2/doc/results.dox [hunk 8]
```diff
@@ -1,14 +1,43 @@
...
+and once after renumbering), along with one mesh file.
+
+Let us start with the DoF locations. There is no particularly convenient
+program to visualize this kind of information, but we can resort to
+<a href="http://www.gnuplot.info/">GNUPLOT</a> (one of the simpler visualization
+programs; maybe not the easiest to use since it is command line driven, but
+also universally available on all Linux and other Unix-like systems). The
+command that produces the following pictures reads as follows:
+```
+plot [-0.1:2.1][-1.1:1.1] "mesh.gnuplot" with lines, "dof-locations-1.gnuplot" using 1:2:3 with labels point offset .3,.2 font "4,6"
```
Comment (bangerth) [inline examples/step-2/doc/results.dox:14]: @tjhei This is where the command to create the picture is. The documentation of the function that creates the data for these figures also links to this location.

**Comment by bangerth** [comment]
Thanks, @tjhei . I've made the modifications you requested almost verbatim.

File: examples/step-2/doc/results.dox (hunk 8 above)
Comment (tjhei) [reply examples/step-2/doc/results.dox:14]: Excellent. How do add a ``set size ratio -1`` or whatever it is called?

Comment (bangerth) [reply examples/step-2/doc/results.dox:14]: I don't know. What do you want to do? I don't know that command...

Comment (tjhei) [reply examples/step-2/doc/results.dox:14]: so that the inside of the mesh is shown with the correct aspect ratio (as a circle).

**Comment by tjhei** [comment]
I will merge for now after blocking your changes for such a long time. We can always do follow-ups...

File: examples/step-2/doc/results.dox (hunk 8 above)
Comment (bangerth) [reply examples/step-2/doc/results.dox:14]: Ah, nice! I didn't know about that command. I've updated the figures on the website now.
